From d7160b0ea2e17741931f415adb1017cc96c42ed1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michel=20D=C3=A4nzer?= <mdaenzer@redhat.com>
Date: Wed, 24 Sep 2025 12:32:22 +0200
Subject: [PATCH 1/2] xwayland: Add heuristic for WM windows based on
 reparenting

If the WM client reparents a window, mark the new parent as a WM window.

This helps with current mutter, where decoration windows are created by
a separate mutter-x11-frames client instead of the WM client. There
might be other compositors doing something similar now or in the future.
---
 hw/xwayland/xwayland-screen.c |  3 +++
 hw/xwayland/xwayland-screen.h |  1 +
 hw/xwayland/xwayland-window.c | 36 ++++++++++++++++++++++++++++++++++-
 hw/xwayland/xwayland-window.h |  1 +
 4 files changed, 40 insertions(+), 1 deletion(-)

diff --git a/hw/xwayland/xwayland-screen.c b/hw/xwayland/xwayland-screen.c
index 59554b7345..22c500801a 100644
--- a/hw/xwayland/xwayland-screen.c
+++ b/hw/xwayland/xwayland-screen.c
@@ -1152,6 +1152,9 @@ xwl_screen_init(ScreenPtr pScreen, int argc, char **argv)
     xwl_screen->ConfigNotify = pScreen->ConfigNotify;
     pScreen->ConfigNotify = xwl_config_notify;
 
+    xwl_screen->ReparentWindow = pScreen->ReparentWindow;
+    pScreen->ReparentWindow = xwl_reparent_window;
+
     xwl_screen->ResizeWindow = pScreen->ResizeWindow;
     pScreen->ResizeWindow = xwl_resize_window;
 
diff --git a/hw/xwayland/xwayland-screen.h b/hw/xwayland/xwayland-screen.h
index 49a167a121..ffbaa09e70 100644
--- a/hw/xwayland/xwayland-screen.h
+++ b/hw/xwayland/xwayland-screen.h
@@ -78,6 +78,7 @@ struct xwl_screen {
     XYToWindowProcPtr XYToWindow;
     SetWindowPixmapProcPtr SetWindowPixmap;
     ChangeWindowAttributesProcPtr ChangeWindowAttributes;
+    ReparentWindowProcPtr ReparentWindow;
     ResizeWindowProcPtr ResizeWindow;
     MoveWindowProcPtr MoveWindow;
     SourceValidateProcPtr SourceValidate;

diff --git a/hw/xwayland/xwayland-window.h b/hw/xwayland/xwayland-window.h
index ba5c3a078b..ae3e531c6e 100644
--- a/hw/xwayland/xwayland-window.h
+++ b/hw/xwayland/xwayland-window.h
@@ -138,6 +138,7 @@ int xwl_config_notify(WindowPtr window,
                       int x, int y,
                       int width, int height, int bw,
                       WindowPtr sib);
+void xwl_reparent_window(WindowPtr window, WindowPtr prior_parent);
 void xwl_resize_window(WindowPtr window,
                        int x, int y,
                        unsigned int width, unsigned int height,
-- 
GitLab
